rustls/tests/api.rs

```diff
@@ -6882,6 +6882,187 @@ impl<'a> io::Write for FakeStream<'a> {
+    server_config.enable_secret_extraction = true;
+
+    let (mut client, mut server) = make_pair_for_configs(client_config, server_config);
+    do_handshake(&mut client, &mut server);
+
+    let mut raw_server = RawTls::new_server(server);
+
+    let msg = PlainMessage {
+        typ: ContentType::Handshake,
+        version: ProtocolVersion::TLSv1_3,
+        payload: Payload::new(vec![
+            u8::from(HandshakeType::HelloRequest),
+            0x00,
+            0x00,
+            0x00,
+        ]),
+    };
+    raw_server.encrypt_and_send(&msg, &mut client);
+    let err = client
+        .process_new_packets()
+        .unwrap_err();
+    assert_eq!(format!("{err:?}"),
+               "InappropriateHandshakeMessage { expect_types: [NewSessionTicket, KeyUpdate], got_type: HelloRequest }");
+}
+
+#[cfg(feature = "tls12")]
+#[test]
+fn test_illegal_server_renegotiation_attempt_after_tls12_handshake() {
+    let client_config =
+        make_client_config_with_versions(KeyType::Rsa2048, &[&rustls::version::TLS12]);
+    let mut server_config = make_server_config(KeyType::Rsa2048);
+    server_config.enable_secret_extraction = true;
+
+    let (mut client, mut server) = make_pair_for_configs(client_config, server_config);
+    do_handshake(&mut client, &mut server);
+
+    let mut raw_server = RawTls::new_server(server);
+
+    let msg = PlainMessage {
+        typ: ContentType::Handshake,
+        version: ProtocolVersion::TLSv1_3,
+        payload: Payload::new(vec![
+            u8::from(HandshakeType::HelloRequest),
+            0x00,
+            0x00,
+            0x00,
+        ]),
+    };
+
+    // one is allowed (and elicits a warning alert)
+    raw_server.encrypt_and_send(&msg, &mut client);
+    client.process_new_packets().unwrap();
+    raw_server.receive_and_decrypt(&mut client, |m| {
+        assert_eq!(format!("{m:?}"),
+                   "Message { version: TLSv1_2, payload: Alert(AlertMessagePayload { level: Warning, description: NoRenegotiation }) }");
+    });
+
+    // second is fatal
+    raw_server.encrypt_and_send(&msg, &mut client);
+    assert_eq!(
+        client
+            .process_new_packets()
+            .unwrap_err(),
+        Error::PeerMisbehaved(PeerMisbehaved::TooManyRenegotiationRequests)
+    );
+}
+
+#[test]
+fn test_illegal_client_renegotiation_attempt_after_tls13_handshake() {
+    let mut client_config =
+        make_client_config_with_versions(KeyType::Rsa2048, &[&rustls::version::TLS13]);
+    client_config.enable_secret_extraction = true;
+    let server_config = make_server_config(KeyType::Rsa2048);
+
+    let (mut client, mut server) = make_pair_for_configs(client_config, server_config);
+    do_handshake(&mut client, &mut server);
+
+    let mut raw_client = RawTls::new_client(client);
+
+    let msg = PlainMessage {
+        typ: ContentType::Handshake,
+        version: ProtocolVersion::TLSv1_3,
+        payload: Payload::new(vec![
+            u8::from(HandshakeType::ClientHello),
```

Comment:
Is there a reason to prefer bashing this together with hex constants instead of building up a `HandshakeMessagePayload` with a `HandshakePayload::ClientHello(ClientHelloPayload)`? It seems like it would be easier to reason about if workable.